### PR TITLE
Fixes issues with empty response bodies

### LINKED ISSUE #46: How to decode a 204?
Hi Guys,

This is a question which I hope you can help me with.

With a 204 response there is no response body to decode, as far as I can tell the current code always requires a decoder, therefor assumes there is always a body and JSON.parse will fail. https://github.com/gcanti/elm-ts/blob/master/src/Http.ts#L153

Is there a NoOp decoder or something?

At the moment I'm just getting the server to return an empty object.

Any help would be fantastic. I need to be able to handle spec compliant responses.

### PR DESCRIPTION
**Pull request by bmazzarol**
Fixes issue with empty response bodies, now they are converted to `{}`

fixes #46

### CODE REVIEW

**Comment by StefanoMagrassi** [comment]
@bmazzarol thank you.

I think that it would be better to split the changes into 2 separated PRs: one for the fix of empty response body and one for the full response headers + `Response` object in `send()` result.

I'll add some comments

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,5 +3,10 @@
   "prettier.printWidth": 120,
   "prettier.semi": false,
   "prettier.singleQuote": true,
-  "tslint.enable": true
+  "tslint.enable": true,
+  "[typescript]": {
```
Comment (StefanoMagrassi) [inline .vscode/settings.json:7]: @bmazzarol please do not change vscode settings: we try to add as less code as possible

File: src/Http.ts [hunk 2]
```diff
@@ -85,6 +85,15 @@ export function send<A, Msg>(f: (e: Either<HttpError, A>) => Msg): (req: Request
   return req => xhr(req).pipe(map(result => T.of(O.some(f(result)))))
 }
 
+/**
+ * Executes as `Cmd` the provided call to remote resource, mapping result to a `Msg` containing the full response.
+ * @category utils
+ * @since 0.5.0
+ */
+export function sendWithResp<A, Msg>(f: (e: Either<HttpError, Response<A>>) => Msg): (req: Request<A>) => Cmd<Msg> {
```
Comment (StefanoMagrassi) [inline src/Http.ts:93]: @bmazzarol this has to be tested (like every other exported/exposed function)

**Comment by bmazzarol** [comment]
Sure @StefanoMagrassi, just let me know what you want and I can split it

File: src/Http.ts [hunk 3]
```diff
@@ -118,13 +127,17 @@ export function post<A>(url: string, body: unknown, decoder: Decoder<A>): Reques
 }
 
 // --- Helpers
-function xhr<A>(req: Request<A>): Observable<Either<HttpError, A>> {
+function xhrWithResp<A>(req: Request<A>): Observable<Either<HttpError, Response<A>>> {
```
Comment (StefanoMagrassi) [inline src/Http.ts:130]: @bmazzarol I think that we can keep only the `xhr` function with a different signature:
```ts
function xhr<A>(req: Request<A>): Observable<Either<HttpError, Response<A>>>
```

and then handle the right side differently in `send()` and `sendWithResp()`:
```ts
type XHR<A> = Observable<Either<HttpError, Response<A>>>

export function toTask<A>(req: Request<A>): TaskEither<HttpError, A> {
  return () => toResult(xhr(req)).toPromise()
}

export function send<A, Msg>(f: (e: Either<HttpError, A>) => Msg): (req: Request<A>) => Cmd<Msg> {
  return req => toResult(xhr(req)).pipe(map(result => T.of(O.some(f(result)))))
}

export function sendWithResp<A, Msg>(f: (e: Either<HttpError, Response<A>>) => Msg): (req: Request<A>) => Cmd<Msg> {
  return req => xhr(req).pipe(map(result => T.of(O.some(f(result)))))
}

// --- Helpers
function toResult<A>(x: XHR<A>): Observable<Either<HttpError, A>> {
  return x.pipe(map(E.map(resp => resp.body)))
}

function xhr<A>(req: Request<A>): XHR<A> {
  return ajax(toXHRRequest(req)).pipe(
    map(flow(toResponse(req), decodeWith(req.expect))),
    catchError((e: unknown): XHR<A> => of(E.left(toHttpError(req, e))))
  )
}

function decodeWith<A>(decoder: Decoder<A>): (resp: Response<string>) => Either<HttpError, Response<A>> {
  return resp =>
    pipe(
      // By spec parsing json can only throw `SyntaxError`
      E.parseJSON(resp.body, e => (e as SyntaxError).message),
      E.chain(decoder),
      E.bimap(
        e => ({
          _tag: 'BadPayload',
          value: e,
          response: resp
        }),
        body => ({ ...resp, body })
      )
    )
}
```

File: src/Http.ts [hunk 4]
```diff
@@ -137,21 +150,29 @@ function toXHRRequest<A>(req: Request<A>): AjaxRequest {
   }
 }
 
+function parseResponseHeaders(rawRespHeaders: string): Record<string, string> {
```
Comment (StefanoMagrassi) [inline src/Http.ts:153]: @bmazzarol I think we can add a field to the `Response` object in order to provide access to the original `XMLHttpRequest` object:

```ts
export interface Response<Body> {
  url: string
  status: {
    code: number
    message: string
  }
  headers: Record<string, string>
  body: Body,
  xhr: XMLHttpRequest
}
```

so every user can choose how to handle it (and we do not have to manage extra code in `elm-ts` source)

File: .vscode/settings.json (hunk 1 above)
Comment (bmazzarol) [reply .vscode/settings.json:7]: Sure mate.

I was having issues with vscode automatically re-ordering all the imports, but not in the order that the linter was happy with. Might have been my vscode setup, but thought it might have been useful for other committers.

Will remove.

File: src/Http.ts (hunk 2 above)
Comment (bmazzarol) [reply src/Http.ts:93]: No problems will copy over the send tests and give it a spin.

File: src/Http.ts (hunk 3 above)
Comment (bmazzarol) [reply src/Http.ts:130]: Sure will do.

File: src/Http.ts (hunk 4 above)
Comment (bmazzarol) [reply src/Http.ts:153]: Yep this will solve all my issues, will do.

**Comment by bmazzarol** [comment]
@StefanoMagrassi let me know if you want any more changes.

File: src/Http.ts [hunk 5]
```diff
@@ -55,6 +55,7 @@ export interface Response<Body> {
   }
   headers: Record<string, string>
   body: Body
+  xhr: XMLHttpRequest
```
Comment (StefanoMagrassi) [inline src/Http.ts:58]: @bmazzarol @gcanti pointed out to me that this kind of change in `Response` type would be **breaking**, thus we could:
1. add it and bump the package's version to **0.6.x** (taking advantage of current [_pre-1.x_ version](https://semver.org/#spec-item-4))
2. do not change the `Response` interface, add the `response` object to the `XHR` type right side and use it `sendWithResp()`
```ts
interface ResponseAndXHR<A> {
  xhr: XMLHttpRequest
  response: Response<A>
}

type XHR<A> = Observable<Either<HttpError, ResponseAndXHR<A>>>

/**
 * Executes as `Cmd` the provided call to remote resource, mapping result to a `Msg` containing the full response and the original XMLHTTPRequest object.
 * @category utils
 * @since 0.5.8
 */
export function sendFull<A, Msg>(f: (e: Either<HttpError, ResponseAndXHR<A>>) => Msg): (req: Request<A>) => Cmd<Msg> {
  return req => xhr(req).pipe(map(result => T.of(O.some(f(result)))))
}

// --- Helpers
function toResult<A>(x: XHR<A>): Observable<Either<HttpError, A>> {
  return x.pipe(map(E.map(({ response }) => response.body)))
}

function xhr<A>(req: Request<A>): XHR<A> {
  return ajax(toXHRRequest(req)).pipe(
    map(aresp =>
      pipe(
        aresp,
        toResponse(req),
        decodeWith(req.expect),
        E.map(response => ({ response, xhr: aresp.xhr })) // <--- here
      )
    ),
    catchError((e: unknown): XHR<A> => of(E.left(toHttpError(req, e))))
  )
}
```
Second option is my favorite...

BTW I would manage this in a different PR

File: test/Http.test.ts [hunk 6]
```diff
@@ -218,7 +221,78 @@ describe('Http', () => {
...
+              _tag: 'BadStatus',
+              response: {
+                url: 'http://example.com/test',
+                status: { code: 500, message: '' },
+                headers: {},
+                body,
+                xhr: server.requests[0]
               }
             }
           })
```
Comment (StefanoMagrassi) [inline test/Http.test.ts:298]: @bmazzarol please add a test also for empty string responses

File: src/Http.ts (hunk 5 above)
Comment (bmazzarol) [reply src/Http.ts:58]: Will create a new pr around the second option.

File: test/Http.test.ts (hunk 6 above)
Comment (bmazzarol) [reply test/Http.test.ts:298]: Sure mate will do.

**Comment by bmazzarol** [comment]
@StefanoMagrassi this PR now should only have the empty response fix, will create another one for the other changes.
